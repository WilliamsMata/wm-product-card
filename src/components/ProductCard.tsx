import React, { CSSProperties, FC, createContext } from 'react';

import { useProduct } from '../hooks/useProduct';
import {
  InitialValues,
  Product,
  ProductCardHandlers,
  ProductContextProps,
  onChangeArgs,
} from '../interfaces/interfaces';

import styles from '../styles/styles.module.css';

export const ProductContext = createContext({} as ProductContextProps);
const { Provider } = ProductContext;

export interface Props {
  children: (args: ProductCardHandlers) => React.ReactNode;
  className?: string;
  product: Product;
  style?: CSSProperties;
  onChange?: (args: onChangeArgs) => void;
  value?: number;
  initialValues?: InitialValues;
}

export const ProductCard: FC<Props> = ({
  children,
  product,
  className,
  style,
  onChange,
  value,
  initialValues,
}) => {
  const {
    counter,
    increaseBy,
    maxCount,
    isMaxCountReached,
    reset,
  } = useProduct({
    product,
    onChange,
    value,
    initialValues,
  });

  return (
    <Provider value={{ counter, increaseBy, product, maxCount }}>
      <div className={`${styles.productCard} ${className}`} style={style}>
        {children({
          count: counter,
          isMaxCountReached,
          maxCount: initialValues?.maxCount,
          product,
          increaseBy,
          reset,
        })}
      </div>
    </Provider>
  );
};
