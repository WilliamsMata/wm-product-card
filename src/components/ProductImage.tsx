import React, { CSSProperties, FC, useContext } from 'react';
import { ProductContext } from './ProductCard';

import styles from '../styles/styles.module.css';
import noImage from '../assets/no-image.jpg';

export interface Props {
  className?: string;
  img?: string;
  style?: CSSProperties;
}

export const ProductImage: FC<Props> = ({ img, className, style }) => {
  const { product } = useContext(ProductContext);

  let imgToShow: string;

  if (img) {
    imgToShow = img;
  } else if (product.img) {
    imgToShow = product.img;
  } else {
    imgToShow = noImage;
  }

  return (
    <img
      src={imgToShow}
      alt="Product"
      className={`${styles.productImg} ${className}`}
      style={style}
    />
  );
};

ProductImage.defaultProps = {
  img: '',
};
