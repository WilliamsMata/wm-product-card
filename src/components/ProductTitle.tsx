import React, { CSSProperties, FC, useContext } from 'react';
import { ProductContext } from './ProductCard';

import styles from '../styles/styles.module.css';

export interface Props {
  className?: string;
  title?: string;
  style?: CSSProperties;
}

export const ProductTitle: FC<Props> = ({ title, className, style }) => {
  const { product } = useContext(ProductContext);

  return (
    <span className={`${styles.productDescription} ${className}`} style={style}>
      {title ? title : product.title}
    </span>
  );
};
