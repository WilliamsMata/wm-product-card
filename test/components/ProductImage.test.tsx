import React from 'react';
import renderer from 'react-test-renderer';
import { ProductImage, ProductCard } from '../../src/components';
import { product2 } from '../data/products';

describe('ProductImage', () => {
  test('Debe de mostrar el componente correctamente la imagen personalizada', () => {
    const imageUrl = './some-image-test.jpg';

    const wrapper = renderer.create(
      <ProductImage img={imageUrl} className="custom-class" />
    );

    const { props } = wrapper.toJSON() as { props: { src?: string } };

    expect(props.src).toEqual(imageUrl);
    expect(wrapper.toJSON()).toMatchSnapshot();
  });

  test('Debe de mostrar el componente la imagen del producto', () => {
    const wrapper = renderer.create(
      <ProductCard product={product2}>{() => <ProductImage />}</ProductCard>
    );

    expect(wrapper.toJSON()).toMatchSnapshot();
  });
});
