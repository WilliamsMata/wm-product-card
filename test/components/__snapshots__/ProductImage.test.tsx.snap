// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductImage Debe de mostrar el componente correctamente la imagen personalizada 1`] = `
<img
  alt="Product"
  className="productImg custom-class"
  src="./some-image-test.jpg"
/>
`;

exports[`ProductImage Debe de mostrar el componente la imagen del producto 1`] = `
<div
  className="productCard undefined"
>
  <img
    alt="Product"
    className="productImg undefined"
    src="./coffee-mug.png"
  />
</div>
`;
